### DISCUSSION TITLE
Can we store all the disposables in a top-level array or set?

### DISCUSSION THREAD

**Discussion by kermanx** [microsoft/vscode-discussions#1195] [Extension Development QnA] [answered]
I saw lots of extensions managing disposables by storing them in class member and dispose them in the class's `dispose` method, for example: 

https://github.com/microsoft/vscode-pull-request-github/blob/b0e67756f4367f389b7d3a301e8523f35cadf5f9/src/gitProviders/vslshost.ts#L82

Or append disposables to `extensionContext.subscriptions`.

I am wondering why not keep a top-level array or set for disposables and dispose all of them in the `deactivate` method like the following code?

```ts
export const disposables: vscode.Disposable[] = []

class SomeView{
  initialize() {
    disposables.push(vscode.window.onXXX)
  }
}

export function activate() {
  new SomeView()
}

export function deactivate() {
  disposables.map(d => d.dispose)
}
```

This will be much simpler, because we don't have to make `SomeView` extend `vscode.disposable`

**Comment by isidorn** [comment, marked as answer]
That is a simplified solution to the problem that is not optimal. You should dispose objects as soon as you no longer need them (for example your view is no longer visible).
Your proposed solution - you would only dispose once VS Code is getting shut down (your extension is getting deactivated). This is not optimal and I highly recommend that you dispose your objects as soon as you no longer need them. Thanks

**Reply by kermanx** [reply]
Thanks for your reply! This makes things much clearer. I highly agree that objects should be disposed once they are no longer needed.

However, I think most of the disposable objects are only disposed when the extension is deactivated. For example, output channel and tree views, and their child-disposables. Is it necessary to manage these disposables in a (traditional) tree structure (i.e. a tree view stores disposables created by it and dispose them when itself is disposed) instead of storing them in a global array?
